<a href="https://colab.research.google.com/github/twumasimb/Exploratory-Data-Analysis/blob/main/EDA_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
# Importing necessary libraries
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import pandas as pd
import statsmodels.api as sm
from scipy.stats import norm
import random

In [19]:
data = pd.read_csv('/content/sample_data/dataset.csv') # Importing the dataset

In [21]:
# Population statistics
data.describe()

ptr  Mean_rating
count  1020.000000  1020.000000
mean     12.724380     3.204280
std       2.493322     0.539454
min       4.070981     1.750000
25%      11.012754     2.875000
50%      12.769690     3.250000
75%      14.688600     3.625000
max      21.200001     4.000000

In [20]:
data = data.drop(['TOT_ENR_M', 'TOT_ENR_F', 'TOT_ENR'], axis=1)

> # **Question 1**




In [ ]:
#randomly selecting 100 datapoints
random.seed(20) # Ensure we get the same results all the time

sampled_data = data.sample(n=100)


In [ ]:
#peeping at sampled data
sampled_data.describe()

ptr  Mean_rating
count  100.000000   100.000000
mean    12.685143     3.212500
std      2.144150     0.552628
min      8.299595     1.750000
25%     10.915094     2.875000
50%     12.598289     3.250000
75%     14.356026     3.625000
max     17.341938     4.000000

In [ ]:
dataset = sampled_data

In [ ]:
dataset.describe() # Sampled data statistics

ptr  Mean_rating
count  100.000000   100.000000
mean    12.685143     3.212500
std      2.144150     0.552628
min      8.299595     1.750000
25%     10.915094     2.875000
50%     12.598289     3.250000
75%     14.356026     3.625000
max     17.341938     4.000000

**Making Statistical Claims**



1.   50% of the ISDs have a student to teacher ratio above 12
2.   The mean student to teacher ratio is at least 12.50



1. ***Sample Proportion***

H0: p = 0.5\
H0: p!= 0.5

In [ ]:
# Count number of ISD with ptr > 12
count = (dataset.ptr > 12).sum().sum()

In [ ]:
count

61

There are 62 ISD with ptr greater than 12. 

Therefore, the sample proportion p_hat = 0.61

In [ ]:
# set the sample size and number of successes
n = 100
p = 0.5
x = 61

# calculate the sample proportion
sample_prop = x / n

# calculate the population proportion
pop_prop = p

# calculate the standard error
se = np.sqrt((pop_prop * (1 - pop_prop)) / n)

# calculate the z-score
z_score = (sample_prop - pop_prop) / se

# calculate the p-value (two tailed hence the factor of 2)
p_value = 2 * (1 - norm.cdf(abs(z_score)))

# print the results
print('z-score:', z_score)
print('p-value:', p_value)


z-score: 2.1999999999999997
p-value: 0.02780689502699718


In [ ]:
# Computing the Confidence Interval

# z-value for 95% confidence level
z = 1.96

#proportions
n = 100
p_hat = 0.61
q_hat = 1 - p_hat

# Error 
E = z*(np.sqrt((p_hat*q_hat)/n))

# Printing the CI 
lower_bound = p_hat - E
upper_bound = p_hat + E

print(f"CI: ({lower_bound}, {upper_bound})")


CI: (0.5144010125576636, 0.7055989874423364)


**Conclusion for Proportion**

The test statistics are\
z-score: 2.1999999999999997\
p-value: 0.02780689502699718\
Hence it is safe to say that we reject the null hypothesis since the p-value is far less than the significant level of alpha = 0.05 or 95%

CI: (0.5144010125576636, 0.7055989874423364)

We are 95% confident that the proportion of ISD with student to teacher ratio above 12 falls within the confidence interval.

2. ***Sample Mean***

H0: mean = 12.50\
H1: mean < 12.50\

Known data about the population\
population mean = 12.72\
population std = 2.49\
sample std = 2.14

In [ ]:
# set the population mean and standard deviation
pop_mean = 12.72
pop_std = 2.49
n = 100
sample_mean = 12.50

# calculate the z-score
z_score = (sample_mean - pop_mean) / (pop_std / np.sqrt(n))

# calculate the p-value (The value is left tailed)
p_value = (1 - norm.cdf(abs(z_score)))

# print the results
print('z-score:', z_score)
print('p-value:', p_value)


z-score: -0.8835341365461872
p-value: 0.18847387353552802


In [ ]:
# Computing the Confidence Interval

# z-value for 95% confidence level
z = 1.96

# mean values
n = 100
sample_mean = 12.50
pop_std = 2.49

# Error 
E = z*(pop_std/np.sqrt(n))

# Printing the CI 
lower_bound = sample_mean - E
upper_bound = sample_mean + E

print(f"CI: ({lower_bound}, {upper_bound})")


CI: (12.01196, 12.98804)


**Conclusion Mean Statistics**

The test statistics are \
z-score: -0.8835341365461872\
p-value: 0.18847387353552802

Hence we fail to reject the null hypothesis since p-value > alpha = 0.05 (95%)

CI: (12.01196, 12.98804)
We are 95% confident that the mean of ISD with student to teacher ratio of 12.50 falls within the confidence interval.

> # ** Question 2**

In [28]:
# randomly selecting 200 datapoints
random.seed(30)

sec_sample = data.sample(n=200)

In [29]:
sec_sample.describe()

ptr  Mean_rating
count  200.000000   200.000000
mean    12.733713     3.190268
std      2.411932     0.549828
min      5.384299     2.000000
25%     11.317993     2.875000
50%     12.836412     3.125000
75%     14.391769     3.625000
max     21.200001     4.000000

***Hypothesis Testing***\
The proportion of ISDs with ptr > 12

H0: p_hat_1 = p_hat_2\
H1: p_hat_1 != p_hat_2

In [30]:
# Count number of ISD with ptr > 12
count = (sec_sample.ptr > 12).sum().sum()
count

131

In [31]:
p_hat_2 = count/sec_sample.count()
p_hat_2

ptr            0.655
Mean_rating    0.655
dtype: float64

In [33]:
x1 = 61
x2 = 131
n1 = 100
n2 = 200

p_bar = (x1+x2)/(n1+n2)
q_bar = 1 - p_bar

print("p_bar: ",p_bar)
print("q_bar: ",q_bar)

p_hat_1 = 0.61
p_hat_2 = 0.655
q_hat_1 = 1 - p_hat_1
q_hat_2 = 1 - p_hat_2

print("p_hat_1: ",p_hat_1)
print("p_hat_2: ",p_hat_2)
print("q_hat_1: ",q_hat_1)
print("q_hat_2: ",q_hat_2)

# Compute the Z-score
z_score = (p_hat_1 - p_hat_2)/np.sqrt(((p_bar*q_bar)/n1)+((p_bar*q_bar)/n2))

# calculate the p-value (multiplying with 2 because it is two tailed)
p_value = 2*(1 - norm.cdf(abs(z_score)))

# print the results
print('z-score:', z_score)
print('p-value:', p_value)

# Computing the confidence interval(confidence of 95%, z_value = 1.96)
z_value = 1.96
Error = z_value*(np.sqrt(((p_hat_1*q_hat_1)/n1) + (p_hat_2*q_hat_2)/n2))

# Printing the CI 
lower_bound = (p_hat_1 - p_hat_2) - Error
upper_bound = (p_hat_1 - p_hat_2) + Error

print(f"CI: ({lower_bound}, {upper_bound})")



p_bar:  0.64
q_bar:  0.36
p_hat_1:  0.61
p_hat_2:  0.655
q_hat_1:  0.39
q_hat_2:  0.345
z-score: -0.7654655446197438
p-value: 0.4439943760230092
CI: (-0.16110208525259145, 0.07110208525259136)


p_bar:  0.64\
q_bar:  0.36\
p_hat_1:  0.61\
p_hat_2:  0.655\
q_hat_1:  0.39\
q_hat_2:  0.345\
z-score: -0.7654655446197438\
p-value: 0.4439943760230092\
CI: (-0.16110208525259145, 0.07110208525259136)\

In [36]:
from scipy.stats import t

x_bar_1 = 12.69
x_bar_2 = 12.73
n1 = 100
n2 = 200

std_1 = 2.14
std_2 = 2.41

p_bar = (x1+x2)/(n1+n2)
q_bar = 1 - p_bar

print("p_bar: ",p_bar)
print("q_bar: ",q_bar)

p_hat_1 = 0.61
p_hat_2 = 0.655
q_hat_1 = 1 - p_hat_1
q_hat_2 = 1 - p_hat_2

print("p_hat_1: ",p_hat_1)
print("p_hat_2: ",p_hat_2)
print("q_hat_1: ",q_hat_1)
print("q_hat_2: ",q_hat_2)

# Calculate the t-value
t_value = (x_bar_1 - x_bar_2) / (np.sqrt((std_1**2)/n1 + (std_2**2)/n2))

# degrees of freedom
df = n1 + n2 - 2 
print("df", df)

# two-tailed test and p-value
p_value = 2 * t.cdf(-np.abs(t_value), df)  

print("t-value: {:.5f}".format(t_value))
print("p-value: {:.5f}".format(p_value))

# Computing the confidence interval(confidence of 95%, t_value = 1.980 (@df = 120))
t_value = 1.980

# Calculate the pooled standard deviation
s_pooled = np.sqrt(((n1-1)*std_1**2 + (n2-1)*std_2**2) / (n1+n2-2))

Error = t_value*(s_pooled * np.sqrt(1/n1 + 1/n2))
print("Error margin:", Error)

# Printing the CI 
lower_bound = (x_bar_1 - x_bar_2) - Error
upper_bound = (x_bar_1 - x_bar_2) + Error

print(f"CI: ({lower_bound}, {upper_bound})")



p_bar:  0.64
q_bar:  0.36
p_hat_1:  0.61
p_hat_2:  0.655
q_hat_1:  0.39
q_hat_2:  0.345
df 298
t-value: -0.14622
p-value: 0.88385
Error margin: 0.5635165308277188
CI: (-0.6035165308277197, 0.5235165308277179)


p_bar:  0.64\
q_bar:  0.36\
p_hat_1:  0.61\
p_hat_2:  0.655\
q_hat_1:  0.39\
q_hat_2:  0.345\
df 298\
t-value: -0.14622\
p-value: 0.88385\
Error margin: 0.5635165308277188\
CI: (-0.6035165308277197, 0.5235165308277179)\